services/QuillLMS/spec/models/classroom_spec.rb

```diff
@@ -200,16 +200,56 @@ def invalid_classroom_attributes
   describe '#hide_appropriate_classroom_units' do
     let(:classroom) { create(:classroom) }
 
-    it 'should call hide_all_classroom_units if classroom not visible' do
-      classroom.visible = false
-      expect(classroom).to receive(:hide_all_classroom_units)
-      classroom.hide_appropriate_classroom_units
+    context 'when called directly' do
+      it 'should call hide_all_classroom_units if classroom not visible and visible changed' do
+        classroom.update!(visible: false) # This saves and sets saved_change_to_visible?
+        expect(classroom).to receive(:hide_all_classroom_units)
+        classroom.hide_appropriate_classroom_units
+      end
+
+      it 'should not call hide_all_classroom_units if classroom not visible but visible did not change' do
+        classroom.update!(visible: false) # First update
+        classroom.reload # Reset saved changes
+        expect(classroom).to_not receive(:hide_all_classroom_units)
+        classroom.hide_appropriate_classroom_units
+      end
+
+      it 'should not call hide_all_classroom_units if classroom visible' do
+        classroom.update!(visible: true)
+        expect(classroom).to_not receive(:hide_all_classroom_units)
+        classroom.hide_appropriate_classroom_units
+      end
     end
 
-    it 'should not do anything if classroom visible' do
-      classroom.visible = true
-      expect(classroom).to_not receive(:hide_all_classroom_units)
-      classroom.hide_appropriate_classroom_units
+    context 'as after_commit callback' do
+      let(:unit) { create(:unit, user: classroom.owner) }
+      let!(:classroom_unit) { create(:classroom_unit, unit:, classroom:, visible: true) }
+
+      it 'should automatically hide classroom units when classroom becomes invisible' do
+        expect(classroom_unit.visible).to be
+
+        classroom.update!(visible: false)
+
+        expect(classroom_unit.reload.visible).not_to be
+      end
+
+      it 'should not hide classroom units when classroom remains visible' do
+        expect(classroom_unit.visible).to be
+
+        classroom.update!(name: 'Updated Name') # Change something other than visible
+
+        expect(classroom_unit.reload.visible).to be
+      end
+
+      it 'should not hide classroom units when classroom becomes visible' do
+        classroom.update!(visible: false) # First make it invisible
+        expect(classroom_unit.reload.visible).not_to be # Verify it was hidden
+
+        classroom_unit.update!(visible: true) # Reset to visible
+        classroom.update!(visible: true) # Now make classroom visible again
+
+        expect(classroom_unit.reload.visible).to be # Should remain visible
```

Comment:
I think this could be
```ruby
expect { classroom.update!(visible: false) }
  .to change { classroom_unit.reload.visible }
  .to(false)

expect {
  classroom_unit.update!(visible: true)
  classroom.update!(visible: true)
}.to change { classroom_unit.reload.visible }
  .to(true)
```